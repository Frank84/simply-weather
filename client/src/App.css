/* */
/* Elements Styling */
/* */

.app {
  z-index: 2;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 100vh;
  background-color: #ddd;
  background-size: cover;
  background-position-y: 50%;
  text-align: center;
  color: white;
}

.app-main {
  z-index: 1;
  position: relative;
  width: auto;
  max-width: 80%;
  margin: 2rem;
  padding: 3rem;
  border-radius: 2rem;
  background-color: rgba(0,0,0,0.5);
  transition: width 0.3s ease-out;
}

.search-wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  position: fixed;
  right: 1rem;
  top: 1rem;
  width: 40px;
  height: 40px;
  padding: 0;
  background: rgba(0,0,0,0.5);
  border-radius: 50%;
  transition: all 0.3s ease-out;
  z-index: 10;
}

.search-wrapper.active {
  right: 0;
  top: 0;
  width: 100%;
  height: 100%;
  background: rgba(0,0,0,0.9);
  border-radius: 0;
}

.search-wrapper form {
  position: relative;
}

.search-wrapper .search-icon {
  position: absolute;
  right: 0;
  top: 40%;
  opacity: 0.5;
}

.search-wrapper .search-icon.hide {
  display: none;
}

.search-wrapper .search-toggle-icon {
  position: relative;
  top: 3px;
  margin: 0;
}

.search-wrapper .search-close {
  position: fixed;
  right: 2rem;
  top: 2rem;
  transition: all 3s ease-out, position 3s ease-out 1s;
  opacity: 0;
}

.search-close.active {
  opacity: 1;
}

.temperature {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  width: 100%;
  margin: auto;
}

.temperature-degree {
  font-size: 5.5rem;
  line-height: 5.5rem;
  text-align: right;
}

.temperature-details {
  display: flex;
  flex-basis: 47%;
  flex-wrap: wrap;
}

.temperature-unit {
  flex-basis: 20%;
  font-size: 2rem;
  color: #999;
  cursor: default;
}

.temperature-unit.active {
  color: #FFF;
}

.temperature-or {
  color: #FFF;
}

.temperature-unit.fahrenheit {
  flex-basis: 70%;
} 

.temperature-feelslike {
  flex-basis: 100%;
  margin: 0.5rem 0;
  padding-left: 0.8rem;
  font-size: 1rem;
  text-align: left;
}

.logo {
  display: flex;
  align-items: center;
  margin: 0;
  font-size: 2rem;
  letter-spacing: -2px;
  color: rgba(255,255,255,1);
  opacity: 0.75;
}

.logo-icon {
  font-size: 2rem;
}

.load {
  border-radius: 2rem;
}

.no-result {
  text-align: left;
  margin: 1rem 0 0 0;
}

.no-result span {
  font-weight: bold;
}

/* */
/* ANIMATIONS */
/* */

@-webkit-keyframes rotating /* Safari and Chrome */ {
  from {
    -webkit-transform: rotate(0deg);
    -o-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  to {
    -webkit-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
@keyframes rotating {
  from {
    -ms-transform: rotate(0deg);
    -moz-transform: rotate(0deg);
    -webkit-transform: rotate(0deg);
    -o-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  to {
    -ms-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
.rotating {
  -webkit-animation: rotating 10s linear infinite;
  -moz-animation: rotating 10s linear infinite;
  -ms-animation: rotating 10s linear infinite;
  -o-animation: rotating 10s linear infinite;
  animation: rotating 10s linear infinite;
}

/* */
/* MEDIA QUERIES */
/* */

@media (min-width: 400px) {
  .temperature-details {
    flex-basis: 45%;
  }
}
@media (min-width: 500px) {
  .temperature-details {
    flex-basis: 40%;
  }
}
@media (min-width: 600px) {
  .app-main {
    padding: 4rem;
  }
  .temperature-degree {
    font-size: 7.2rem;
    line-height: 7.2rem;
  }
  .temperature-details {
    flex-basis: 30%;
  }
  .search-wrapper {
    width: 50px;
    height: 50px;
  } 
  .search-wrapper .icon {
    font-size: 1.8rem;
  }
  .search-wrapper .search-close {
    top: 2rem;
  }
  .app-header .logo .icon {
    font-size: 2.4rem;
  }
  .logo-icon {
    font-size: 2.5rem;
  }
}

@media (min-width: 800px) {
  .temperature-details {
    flex-basis: 20%;
  }
}