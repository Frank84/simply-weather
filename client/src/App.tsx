import geolocator from 'geolocator';
import moment from 'moment';
import React, { useEffect, useState } from 'react';
import { Icon } from 'semantic-ui-react'

import './App.css';

import WeatherDetails from './components/weather-details';
import defaultBackgroundImage from './images/background.png';
import { getRandomInt } from './utilities/utils';

function App() {
  const [currentLocation, setCurrentLocation] = useState('');
  const [currentWeather, setCurrentWeather] = useState({
    current: {},
    historical: {},
    location: { name: '' },
  });
  const [degreeUnit, setDegreeUnit] = useState('c');
  const [isError, setIsError] = useState(false);
  const [loading, setLoading] = useState(false);
  const [locationBackgroundImage, setLocationBackgroundImage] = useState('');
  const [menuOpen, setMenuOpen] = useState(false);
  const [submitValue, setSubmitValue] = useState('');

  async function getWeather(query: string) {
    // Getting weather forecast and historical data using WeatherStack API.
    // Please note that we are using the free version of the API. 
    // For this reason, historical data will be use for the forecast.
    // Parameters passed: 
    // - range: string of the past 30 days.
    // - query: name of the city requested
    // - unit: m (metrics) f (fahrenheit)
    setIsError(false);
    setLoading(true);
    const range = getStringDays(30);
    await fetch(`/weather/historical?t=${range}&q=${query}&unit=${degreeUnit}`)
      .then((response) => {
        if (response.status >= 400 && response.status < 600) {
          setIsError(true);
          throw new Error("Bad response from server");
        }
        return response;
      }).then((returnedResponse) => {
        returnedResponse.json()
          .then((parsedJson) => {
            const results = parsedJson.results;
            setCurrentWeather({ current: results.current, location: results.location, historical: results.historical });
            setCurrentLocation(results.location.name);
          });
      }).catch((error) => {
        setIsError(true);
        console.error(error);
      }).finally(() => {
        setLoading(false);
      });
  }

  async function getBackgroundImage(query: string) {
    // Try to get a background image based on the city selected using Unsplash API.
    // Use Unsplash API. If no result found, or error, use default background image. 
    try {
      const apiQuery = await fetch(`/images?q=${query}`);
      const parsedJson = await apiQuery.json();
      if (parsedJson && parsedJson.results && parsedJson.results.results) {
        const results = parsedJson.results.results;
        const randomIndex = getRandomInt(results.length);
        setLocationBackgroundImage(results[randomIndex].urls.regular);
      } else {
        setLocationBackgroundImage(defaultBackgroundImage);
      }
    }
    catch {
      setLocationBackgroundImage(defaultBackgroundImage);
    }
  }

  async function searchNewLocation(e: React.SyntheticEvent<EventTarget>) {
    e.preventDefault();
    await getWeather(submitValue);
    setMenuOpen(false);
  }

  function getStringDays(numberOfdays: Number): string {
    // This return a string in the format expected by WeatherStack API.
    // Format: 'YYYY-MM-DD;YYYY-MM-DD...'
    let range = '';
    for (let i = 0; i < numberOfdays; i++) {
      let date = moment().subtract(i, 'days').format('YYYY-MM-DD') + ';';
      range = range + date;
    } 
    return range.substring(0, range.length - 1);
  }

  useEffect(() => {
    if (currentWeather && currentWeather.location) {
      // Make sure the location is a different one before getting new picture.
      if (currentWeather.location.name !== currentLocation) {
        getBackgroundImage(currentWeather.location.name)
      }
    }
  }, [currentWeather]);

  useEffect(() => {
    var options = {
        enableHighAccuracy: true,
        timeout: 5000,
        maximumWait: 10000,     // max wait time for desired accuracy
        maximumAge: 0,          // disable cache
        desiredAccuracy: 30,    // meters
        fallbackToIP: true,     // fallback to IP if Geolocation fails or rejected
    };
    geolocator.locate(options, function (err: any, location: any) {
        if (err) return console.log(err);
        if (location && location.address && location.address.city) {
          getWeather(location.address.city);
        } else if (location && location.coords) {
          getWeather(location.coords.latitude + ',' + location.coords.longitude);
        } else {
          setMenuOpen(true);
        }
    });
  }, []);

  function renderNewLocationForm() {
    return (
      <div>
        <form onSubmit={searchNewLocation}>
          <div className={`ui action transparent input ${loading ? 'loading' : ''}`}>
            <input autoFocus type="text" onChange={(e) => setSubmitValue(e.target.value)} placeholder="Search a city" />
            <i className={`search icon`} onClick={searchNewLocation}></i>
          </div>
        </form>
        <Icon className={`search-close ${menuOpen ? 'active' : ''}`} onClick={() => { setMenuOpen(false) }} name='close' />
      </div>
    )
  }

  return (
    <div className="app" style={{backgroundImage: `url(${locationBackgroundImage})`}}>
      <header className="app-header">
        <h1 className="logo">
          <Icon className='logo-icon rotating' name='sun' /> 
          Simply Weather
        </h1>
        <div className={`search-wrapper ${menuOpen ? 'active' : '' }`}>
          { menuOpen ? renderNewLocationForm() : 
            <Icon name='search' onClick={() => { setMenuOpen(true) }} /> 
          }
        </div>
      </header>
      <main className="app-main">
        { !isError ? currentWeather.location.name ?
          <WeatherDetails getWeather={getWeather} currentWeather={currentWeather} degreeUnit={{degreeUnit, setDegreeUnit}} /> :
          <div className="load ui active dimmer"><div className="ui loader"></div></div> :
          <div>Oops, something went wrong. <span style={{ textDecoration: 'underline', cursor: 'default' }} onClick={() => { getWeather(currentLocation) }}>Please try again.</span></div>
        }
      </main>
    </div>
  );
}

export default App;