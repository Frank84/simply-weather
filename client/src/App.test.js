import App from './App';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

it('renders properly', () => {
  const wrapper = shallow(<App />);
  expect(toJson(wrapper)).toMatchSnapshot();
});