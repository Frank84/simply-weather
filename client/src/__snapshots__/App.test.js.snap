// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div
  className="app"
  style={
    Object {
      "backgroundImage": "url()",
    }
  }
>
  <header
    className="app-header"
  >
    <h1
      className="logo"
    >
      <Icon
        as="i"
        className="logo-icon rotating"
        name="sun"
      />
      Simply Weather
    </h1>
    <div
      className="search-wrapper "
    >
      <Icon
        as="i"
        className="search-toggle-icon"
        name="search"
        onClick={[Function]}
      />
    </div>
  </header>
  <div
    className="ui active centered inline loader"
  />
</div>
`;
